Prevent memory leaks in AutoClearedValue

Svechnikov opened the pull request:
When using AutoClearedValue we expect it to clear the value as soon
as the fragment view is destroyed. But in fact, the value is cleared
only when the fragment itself is destroyed, which makes AutoClearedValue
useless as the destroyed fragment is already available to be garbage
collected.

In order to solve the problem we should instead use viewLifecycleOwner.
But it's not available right away, so we wait till fragment lifecycle
ON_START event.

Svechnikov commented:
@googlebot I signed it!

Svechnikov commented:
Some related issues:
https://github.com/android/architecture-components-samples/issues/69
https://github.com/android/architecture-components-samples/issues/673

jeffcp28 commented:
Once the Fragment's view is destroyed (Fragment's onDestroyView()), and then another view created (Fragment's onCreateView()), won't we need to re-register our `viewObserver` on the new view's lifecycle owner?

For example, by setting `observerRegistered` back to false in our `onDestroyView()`, so that the registration logic can run again when the Fragment does onStart().

jeffpenga commented:
When the Fragment's view gets destroyed and then re-created, won't we need to re-register our `viewObserver` onto the new view's lifecycle owner? That way the stored value will be nulled out again the next time this new view also gets destroyed.

For example, by setting `observerRegistered` back to false in our `onDestroyView()`, so that the registration process will happen again (with the new view's lifecycle owner) when the Fragment next does `onStart()`.

Svechnikov commented:
jeffpenga, you are right, thank you. I modified the pull-request. The `observerRegistered` is now reset every time the Fragment's view is destroyed.

IonutNegru87 commented:
I manually integrated your solution as I also suspected the initial version for `AutoClearedValue` leaked memory, but I have encountered some issues regarding the tests.
It seems all the tests fail due to `IllegalStateException` - LifecycleOwner is accessed outside the fragment lifecycle.

Svechnikov commented:
Hello @IonutNegru87. Thank you for reporting the issue with the tests. The problem was that the test fragment didn't return a non-null view in its `onCreateView` method. I fixed the tests and updated my commit. Now the errors are gone.

ianhanniballake commented:
You're totally right about `AutoClearedValue`. Unfortunately, I didn't see this pull request before merging in #785, which covers this and a few other cases.
